viewing_party/party.py

```diff
@@ -1,23 +1,216 @@
 # ------------- WAVE 1 --------------------
+import copy
 
 def create_movie(title, genre, rating):
-    pass
+
+    if title != None and genre != None and rating != None:
+        movie_dict = {"title":title, "genre":genre, "rating":rating}
+        return movie_dict
+    else: 
+        return None
+
+
+
+
+def add_to_watched(user_data, movie):
+    user_data["watched"].append(movie)
+    return user_data
+
+
+
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+    return user_data    
+
+
+
+
+def watch_movie(user_data, title):
+
+    for movie in user_data["watchlist"]:
+        if movie["title"]==title:
+            user_data["watched"].append(movie)
+            user_data["watchlist"].remove(movie)
+
+    
+    return user_data
+
+
+
+
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    avg = 0
+    counter = 0
+    for i in user_data["watched"]:
+        for k,v in i.items():
+            if k == "rating":
+                avg+=v
+                counter+=1
+    if counter!=0:
+        avg = avg/counter
+    return avg
+
+
+def get_most_watched_genre(user_data):
+
+    if len(user_data["watched"]) == 0:
+        return None
+
+    genre_dict = {}
+    for i in user_data["watched"]:
+
+            for k,v in i.items():
+
+                if k == "genre":
+
+                    key_found = False
+                    for key in genre_dict.keys():
+
+                        if key == v:
+
+                            genre_dict[key]+=1
+
+                            key_found = True
+
+                    if key_found == False:
+                        genre_dict[v] = 1
+
+    most_watched_num = 0
+    most_watched_genre = ""
+    for g,f in genre_dict.items():
+
+        if f>most_watched_num:
+            most_watched_genre = g
+            most_watched_num = f
+
+    return most_watched_genre
+        
+
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+
+def get_unique_watched(user_data):
+
+
+    unique_movies = copy.deepcopy(user_data["watched"])

```

Comment:
A shallow copy would also work here.  `user_data['watched'].copy()` will create a new list that points to the same dictionaries.  In line 115, the remove will only remove the dictionaries from the copy of user_data['watched'] and will not make changes to user_data['watched'].  The deep copy is only needed if changes will be made to the dictionaries in the lists and the original version of the dictionaries also need to be preserved.